storm-client/src/jvm/org/apache/storm/utils/ShellLogHandler.java

```diff
@@ -0,0 +1,113 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.storm.utils;
+
+import org.apache.logging.log4j.ThreadContext;
+import org.apache.storm.multilang.ShellMsg;
+import org.apache.storm.task.TopologyContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class ShellLogHandler {
+    public static final Logger LOG = LoggerFactory.getLogger(ShellLogHandler.class);
+
+    public static final String ID = "id";
+    public static final String NAME = "name";
+    public static final String PARENT = "parent";
+    public static final String PID = "pid";
+    public static final String TASK = "task";
+
+    private ShellLogHandler() {
+    }
+
+    private static void putIfNotNull(String key, Object value) {
+        if (value != null) {
+            ThreadContext.put(key, value.toString());
+        }
+    }
+
+    /**
+     * Update the {@link ThreadContext} with information about the logged
+     * message, including the pid and task.
+     * 
+     * @param shellMsg
+     *            - the {@link ShellMsg} containing the ID.
+     * @param process
+     *            - the current {@link ShellProcess}.
+     * @param context
+     *            - the current {@link TopologyContext}.
+     */
+    private static void updateContext(ShellMsg shellMsg, ShellProcess process, TopologyContext context) {
+        putIfNotNull(ID, shellMsg.getId());
+        // Calling this only once allows the same parent ID to be attached to
+        // all log messages from a tuple tree
+        if (!ThreadContext.containsKey(PARENT)) {
+            putIfNotNull(PARENT, shellMsg.getId());
+        }
```

Comment:
This really needs some kind of documentation if we are going to start using id with logging.  I just don't know how you expect to use this, so I have no idea how to say if it works properly or not.

Also you missed javascript support.

https://github.com/apache/storm/blob/a18657d0cd01601dfd193dccade1f601fbef94b4/storm-multilang/javascript/src/main/resources/resources/storm.js#L51-L57

Reply:
Thanks for looking at this @revans2.

Our use case is that we're running a Storm topology written in Python. As well as human-readable logs, the topology is also configured to write JSON logs which are sent to ELK. 

The parent ID is so that we can easily search for all log messages associated with a particular originating tuple.

@HeartSaVioR has already made some comments on the [issue](https://issues.apache.org/jira/browse/STORM-2862), saying that modifying the format of the logs could break backwards compatibility. I've asked him about maybe making the logging output more configurable instead. If we went that way, we could drop the Python and Ruby part of this change, and handle that in our code.

Sorry, I'd explained the missing JS support on the #2474 but it should have been here -  the JS is written to support variable arguments (probably because of [STORM-2435](https://issues.apache.org/jira/browse/STORM-2435)). I can't see a way to support optional IDs without breaking the current behaviour.

Reply:
I think letting users leave the log message with providing enough information is most flexible way. `handleLog` is adding some context which some users might not want and in some case could disturb parsing the message.
So the ideal approach from my side is publishing all the necessary context to multi-lang running subprocess and let subprocess build the full message (without context which log4j2 format provides), but it might bring some backward incompatibility like I said, or even some overhead because it should be passed to multilang and longer message should be passed to ShellSpout/ShellBolt.

I'm OK to follow proposal provided by @hmcc, which may not require huge change. ID of the originating tuple is not available (we're sending generated new ID to multilang) so need to be added in some way though.

Reply:
@hmcc 

If parent id is intended to be used to search for all log messages about a given tuple tree then there is a bug.  You set parent ID once and only once.  This means that "parent" will be set to the ID of the first log message with an ID ever sent.  Every single tuple is likely to have a different parent so it will end up being a totally useless value.


Also I agree with @HeartSaVioR I think there is a much cleaner way to get the same results that you want, and not having all of the overhead for those that don't want it.  Because the logging is thread specific.

I would propose a ShellLogHandler interface that can be set for implementations of ShellBolt and ShellSpout.  That way you can override it, and there is no overhead for people who don't want it.  I would propose something like

```
interface ShellLogHandler {
  void setupContext(ShellProcess process, TopologyContext context);
  void log(ShellMessage msg);
}
```

`setupContext` would be called from the BoltReaderRunnable thread.  It could set the ThreadContext for that thread for things like the PID, etc.  These are things that will not change during the lifetime of the Bolt.

For the spout it is going to be more overhead because there is no way currently to know if you are on a thread by yourself or sharing a thread with another spout.  So for each log message the spout would call `setupContext` followed by `log`.

You also don't need to modify storm.py as you can just use sendMsgToParent yourself directly and have your own custom helper log methods with the ID.

Reply:
Thanks, that sounds like a plan.

What do you prefer - close this and create a new PR, or leave this open, revert the commit and add a new commit with those changes?

Reply:
Either is fine with me.